hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMetadataMergedLogRecordScanner.java

```diff
@@ -0,0 +1,308 @@
+import org.apache.hudi.common.util.HoodieRecordSizeEstimator;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.ReflectionUtils;
+import org.apache.hudi.common.util.SpillableMapUtils;
+import org.apache.hudi.common.util.collection.ExternalSpillableMap;
+import org.apache.hudi.exception.HoodieIOException;
+import org.apache.hudi.internal.schema.InternalSchema;
+import org.apache.hudi.storage.HoodieStorage;
+import org.apache.hudi.storage.StoragePath;
+
+import org.apache.avro.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.apache.hudi.common.fs.FSUtils.getRelativePartitionPath;
+import static org.apache.hudi.common.util.StringUtils.nonEmpty;
+import static org.apache.hudi.common.util.ValidationUtils.checkArgument;
+
+public class HoodieMetadataMergedLogRecordScanner extends HoodieMergedLogRecordScanner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(HoodieMetadataMergedLogRecordScanner.class);
+  // Map of compacted/merged records
+  private final ExternalSpillableMap<Serializable, HoodieRecord> records;
+
+  private HoodieMetadataMergedLogRecordScanner(HoodieStorage storage, String basePath, List<String> logFilePaths, Schema readerSchema,
+                                               String latestInstantTime, Long maxMemorySizeInBytes,
+                                               boolean reverseReader, int bufferSize, String spillableMapBasePath,
+                                               Option<InstantRange> instantRange,
+                                               ExternalSpillableMap.DiskMapType diskMapType,
+                                               boolean isBitCaskDiskMapCompressionEnabled,
+                                               boolean withOperationField, boolean forceFullScan,
+                                               Option<String> partitionName,
+                                               InternalSchema internalSchema,
+                                               Option<String> keyFieldOverride,
+                                               boolean enableOptimizedLogBlocksScan, HoodieRecordMerger recordMerger,
+                                               Option<HoodieTableMetaClient> hoodieTableMetaClientOption) {
+    super(storage, basePath, logFilePaths, readerSchema, latestInstantTime, maxMemorySizeInBytes, reverseReader, bufferSize, spillableMapBasePath, instantRange, diskMapType,
+        isBitCaskDiskMapCompressionEnabled, withOperationField, false, partitionName, internalSchema, keyFieldOverride, enableOptimizedLogBlocksScan, recordMerger,
+        hoodieTableMetaClientOption);
+    try {
+      // Store merged records for all versions for this log file, set the in-memory footprint to maxInMemoryMapSize
```

Comment:
If we make the super class `HoodieMergedLogRecordScanner` records cache key as generic type, we can actually eliminate the redundant initialization of the map.

Reply:
Done.